# Project 1: Organizing Netflix Movie Actor Rating Data

## Analysis Objectives

The purpose of this data analysis is to organize the average IMDB ratings of actors' works in different genres, such as comedy, action, and sci-fi films, in order to identify actors associated with high-rated works in each genre.

## Introduction

The original dataset records all Netflix TV shows and movies available in the United States as of July 2022. The dataset consists of two tables: `titles.csv` and `credits.csv`.

The `titles.csv` file contains information related to movies and TV shows, including IDs, titles, types, descriptions, genres, and IMDB ratings (an international online rating platform), among other details. The `credits.csv` file contains information on over 70,000 directors and actors who appeared in Netflix productions, including names, work IDs, character names, and role types (director/actor).

The meaning of each column in `titles.csv` is as follows:
- **id**: The unique ID of the movie or TV show.
- **title**: The title of the movie or TV show.
- **show_type**: The type of content, either TV show or movie.
- **description**: A brief description of the content.
- **release_year**: The year of release.
- **age_certification**: Age certification rating.
- **runtime**: Length of each episode or the movie in minutes.
- **genres**: A list of genre types.
- **production_countries**: A list of production countries.
- **seasons**: Number of seasons if it is a TV show.
- **imdb_id**: The IMDB ID of the content.
- **imdb_score**: IMDB rating score.
- **imdb_votes**: Number of votes on IMDB.
- **tmdb_popularity**: Popularity on TMDB.
- **tmdb_score**: Rating score on TMDB.

The meaning of each column in `credits.csv` is as follows:
- **person_ID**: The unique ID of the cast or crew member.
- **id**: The ID of the movie or TV show they participated in.
- **name**: The name of the person.
- **character_name**: The name of the character played.
- **role**: The role type, either actor or director.

## Loading Data

Import the necessary libraries for data analysis, and use Pandas' `read_csv` function to parse the contents of the "titles.csv" file into a DataFrame and assign it to the variable `original_titles`. Similarly, parse the contents of the "credits.csv" file into a DataFrame and assign it to the variable `original_credits`.

In [1]:
import pandas as pd

In [2]:
original_titles = pd.read_csv("titles.csv")
original_credits = pd.read_csv("credits.csv")

In [3]:
original_titles.head()

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600

In [4]:
original_credits.head()

person_id       id             name                character   role
0       3748  tm84618   Robert De Niro            Travis Bickle  ACTOR
1      14658  tm84618     Jodie Foster            Iris Steensma  ACTOR
2       7064  tm84618    Albert Brooks                      Tom  ACTOR
3       3739  tm84618    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR
4      48933  tm84618  Cybill Shepherd                    Betsy  ACTOR

## Data Assessment and Cleaning

In this section, we will evaluate and clean the data contained in the `original_titles` and `original_credits` DataFrames established in the previous step.

The process will focus on two key aspects: **structure** and **content**, namely tidiness and cleanliness.

- **Structural issues** refer to problems where the data does not conform to the standard of "each variable as a column, each observation as a row, and each type of observational unit as a cell."
- **Content-related issues** include missing data, duplicate entries, inconsistent data, or invalid data.

To differentiate the cleaned data from the original data, we will create a new variable `cleaned_titles` as a copy of `original_titles`, and another variable `cleaned_credits` as a copy of `original_credits`. All subsequent cleaning steps will be applied to `cleaned_titles` and `cleaned_credits`.

In [5]:
cleaned_titles = original_titles.copy()
cleaned_credits = original_credits.copy()

### Data Tidiness

In [6]:
cleaned_titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   
5   ts22164         Monty Python's Flying Circus   SHOW   
6   tm70993                        Life of Brian  MOVIE   
7   tm14873                          Dirty Harry  MOVIE   
8  tm119281                     Bonnie and Clyde  MOVIE   
9   tm98978                      The Blue Lagoon  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   
5  A British sketch comedy series with the shows ...          1969   
6  Brian Cohen is an average young Jewish man, bu...          1979   
7  When a madman dubbed 'Scorpio' terrorizes San ...          1971   
8  In the 1930s, bored waitress Bonnie Parker fal...          1967   
9  Two small children and a ship's cook survive a...          1980   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   
5             TV-14       30                       ['comedy', 'european']   
6                 R       94                                   ['comedy']   
7                 R      102              ['thriller', 'action', 'crime']   
8                 R      110                 ['crime', 'drama', 'action']   
9                 R      104               ['romance', 'action', 'drama']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   
5               ['GB']      4.0  tt0063929         8.8     73424.0   
6               ['GB']      NaN  tt0079470         8.0    395024.0   
7               ['US']      NaN  tt0066999         7.7    155051.0   
8               ['US']      NaN  tt0061418         7.7    112048.0   
9               ['US']      NaN  tt0080453         5.8     69844.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600  
5           17.617       8.306  
6           17.770       7.800  
7           12.817       7.500  
8           15.687       7.500  
9           50.324       6.156

Based on a sample of 10 rows of data, the `genres` and `production_countries` columns in `cleaned_titles` contain multiple values and should be split.

First, extract a value from the `genres` column for observation.

In [7]:
cleaned_titles['genres'][1]

"['drama', 'crime']"

Although the `genres` column appears to be a list, its actual data type is a string rather than a list of strings, making it impossible to directly use `value_counts` to count the occurrences of each value. 

We can use Python's built-in `eval` function, which converts a string into an expression, helping us transform the string representation of a list into an actual list.

In [8]:
cleaned_titles['genres'] = cleaned_titles['genres'].apply(lambda x: eval(x))
cleaned_titles['genres'][1]

['drama', 'crime']

After converting to a list, we can use the DataFrame's `explode` method to split the list values in that column into individual rows.

In [9]:
cleaned_titles = cleaned_titles.explode("genres")
cleaned_titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   

  age_certification  runtime         genres production_countries  seasons  \
0             TV-MA       51  documentation               ['US']      1.0   
1                 R      114          drama               ['US']      NaN   
1                 R      114          crime               ['US']      NaN   
2                 R      109          drama               ['US']      NaN   
2                 R      109         action               ['US']      NaN   
2                 R      109       thriller               ['US']      NaN   
2                 R      109       european               ['US']      NaN   
3                PG       91        fantasy               ['GB']      NaN   
3                PG       91         action               ['GB']      NaN   
3                PG       91         comedy               ['GB']      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811

Next, apply the same process to the `production_countries` column.

In [10]:
cleaned_titles['production_countries'][1]

1    ['US']
1    ['US']
Name: production_countries, dtype: object

In [11]:
cleaned_titles['production_countries'] = cleaned_titles['production_countries'].apply(lambda s: eval(s))
cleaned_titles['production_countries'][1]

1    [US]
1    [US]
Name: production_countries, dtype: object

In [12]:
cleaned_titles = cleaned_titles.explode('production_countries')
cleaned_titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   

  age_certification  runtime         genres production_countries  seasons  \
0             TV-MA       51  documentation                   US      1.0   
1                 R      114          drama                   US      NaN   
1                 R      114          crime                   US      NaN   
2                 R      109          drama                   US      NaN   
2                 R      109         action                   US      NaN   
2                 R      109       thriller                   US      NaN   
2                 R      109       european                   US      NaN   
3                PG       91        fantasy                   GB      NaN   
3                PG       91         action                   GB      NaN   
3                PG       91         comedy                   GB      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811

After addressing the structural issues in `cleaned_titles`, examine `cleaned_credits`.

In [13]:
cleaned_credits.head(10)

person_id       id             name                      character   role
0       3748  tm84618   Robert De Niro                  Travis Bickle  ACTOR
1      14658  tm84618     Jodie Foster                  Iris Steensma  ACTOR
2       7064  tm84618    Albert Brooks                            Tom  ACTOR
3       3739  tm84618    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR
4      48933  tm84618  Cybill Shepherd                          Betsy  ACTOR
5      32267  tm84618      Peter Boyle                         Wizard  ACTOR
6     519612  tm84618   Leonard Harris      Senator Charles Palantine  ACTOR
7      29068  tm84618   Diahnne Abbott                Concession Girl  ACTOR
8     519613  tm84618      Gino Ardito             Policeman at Rally  ACTOR
9       3308  tm84618  Martin Scorsese  Passenger Watching Silhouette  ACTOR

From the first 10 rows of data, `cleaned_credits` conforms to the principle of "each variable as a column, each observation as a row, and each type of observational unit as a separate cell." Therefore, there are no structural issues.

### Data Cleanliness

Next, use `info` to get an overview of the data content.

In [14]:
cleaned_titles.info()

<class 'pandas.core.frame.DataFrame'>
Index: 17818 entries, 0 to 5849
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    17818 non-null  object 
 1   title                 17817 non-null  object 
 2   type                  17818 non-null  object 
 3   description           17790 non-null  object 
 4   release_year          17818 non-null  int64  
 5   age_certification     10889 non-null  object 
 6   runtime               17818 non-null  int64  
 7   genres                17755 non-null  object 
 8   production_countries  17439 non-null  object 
 9   seasons               6224 non-null   float64
 10  imdb_id               17116 non-null  object 
 11  imdb_score            16976 non-null  float64
 12  imdb_votes            16945 non-null  float64
 13  tmdb_popularity       17663 non-null  float64
 14  tmdb_score            17241 non-null  float64
dtypes: float64(5), int64(2), 

From the output results, the `cleaned_titles` data contains a total of 17,818 observations. The variables `title`, `description`, `age_certification`, `genres`, `production_countries`, `seasons`, `imdb_id`, `imdb_score`, `imdb_votes`, `tmdb_popularity`, and `tmdb_score` all have missing values, which will be assessed and cleaned in the subsequent steps.

Additionally, the `release_year` represents a year and should not be in numeric format; it should be converted to a date format.

In [15]:
cleaned_titles["release_year"] = pd.to_datetime(cleaned_titles["release_year"], format='%Y')
cleaned_titles["release_year"]

0      1945-01-01
1      1976-01-01
1      1976-01-01
2      1972-01-01
2      1972-01-01
          ...    
5847   2021-01-01
5848   2021-01-01
5849   2021-01-01
5849   2021-01-01
5849   2021-01-01
Name: release_year, Length: 17818, dtype: datetime64[ns]

In [16]:
cleaned_credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77801 entries, 0 to 77800
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   person_id  77801 non-null  int64 
 1   id         77801 non-null  object
 2   name       77801 non-null  object
 3   character  68029 non-null  object
 4   role       77801 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.0+ MB


From the output results, the `cleaned_credits` data contains a total of 77,801 observations. The `character` variable has missing values, which will be assessed and cleaned in the subsequent steps.

Additionally, `person_id` represents the cast or crew ID and should not be in numeric format; it should be a string. Therefore, data type conversion is required.

In [17]:
cleaned_credits["person_id"] = cleaned_credits["person_id"].astype(str)
cleaned_credits["person_id"]

0           3748
1          14658
2           7064
3           3739
4          48933
          ...   
77796     736339
77797     399499
77798     373198
77799     378132
77800    1950416
Name: person_id, Length: 77801, dtype: object

#### Handling Missing Data

In `cleaned_titles`, the variables `title`, `description`, `age_certification`, `genres`, `production_countries`, `seasons`, `imdb_id`, `imdb_score`, `imdb_votes`, `tmdb_popularity`, and `tmdb_score` contain missing values.

Since the movie or show title, description, age certification, production countries, seasons, IMDb ID, IMDb votes, TMDb popularity, and TMDb score do not impact our analysis of actors in high IMDb-rated works across different genres, we can retain the observations with missing values in the `title`, `description`, `age_certification`, `production_countries`, `seasons`, `imdb_id`, `imdb_votes`, `tmdb_popularity`, and `tmdb_score` columns.

However, the `imdb_score` and `genres` variables (i.e., IMDb score and genres) are closely related to the analysis we aim to conduct.

Let's first extract and examine the observations with missing `imdb_score` values.

In [18]:
cleaned_titles.query("imdb_score.isnull()")

id                                      title   type  \
0      ts300399        Five Came Back: The Reference Films   SHOW   
75     tm132164                       Bill Hicks: Sane Man  MOVIE   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
...         ...                                        ...    ...   
5810  tm1225897                                 Social Man  MOVIE   
5833   ts307884                                 HQ Barbers   SHOW   
5840  tm1216735                            Sun of the Soil  MOVIE   
5844  tm1074617              Bling Empire - The Afterparty  MOVIE   
5848  tm1035612  Dad Stop Embarrassing Me - The Afterparty  MOVIE   

                                            description release_year  \
0     This collection includes 12 World War II-era p...   1945-01-01   
75    Sane Man was filmed before Bill recorded ‘Dang...   1989-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
...                                                 ...          ...   
5810  Two competitive social media Influencers go he...   2021-01-01   
5833  When a family run barber shop in the heart of ...   2021-01-01   
5840  In 14th-century Mali, an ambitious young royal...   2022-01-01   
5844  The stars of "Bling Empire" discuss the show's...   2021-01-01   
5848  Jamie Foxx, David Alan Grier and more from the...   2021-01-01   

     age_certification  runtime         genres production_countries  seasons  \
0                TV-MA       51  documentation                   US      1.0   
75                   R       80         comedy                   US      NaN   
145               TV-G       18  documentation                   JP     12.0   
145               TV-G       18         family                   JP     12.0   
145               TV-G       18        reality                   JP     12.0   
...                ...      ...            ...                  ...      ...   
5810               NaN       96          drama                  NaN      NaN   
5833             TV-14       24         comedy                   NG      1.0   
5840               NaN       26            NaN                  NaN      NaN   
5844               NaN       35            NaN                   US      NaN   
5848             PG-13       37            NaN                   US      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0            NaN         NaN         NaN            0.600         NaN  
75           NaN         NaN         NaN            3.377         7.5  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
...          ...         ...         ...              ...         ...  
5810  tt20198164         NaN         NaN              NaN         NaN  
5833         NaN         NaN         NaN            0.840         NaN  
5840         NaN         NaN         NaN            1.179         7.0  
5844         NaN         NaN         NaN              NaN         NaN  
5848         NaN         NaN         NaN            1.296        10.0  

[842 rows x 15 columns]

Since the core data needed for analysis, `imdb_score`, is missing, we will remove these observations and then check the number of missing values in this column after deletion:

In [19]:
cleaned_titles = cleaned_titles.dropna(subset=["imdb_score"])
cleaned_titles["imdb_score"].isnull().sum()

0

Next, extract and examine the observations with missing `genres` values.

In [20]:
cleaned_titles.query("genres.isnull()")

id                                              title   type  \
1813    ts77824  My Next Guest Needs No Introduction With David...   SHOW   
1939   ts215037                              Minecraft: Story Mode   SHOW   
2386    ts74805                   A Little Help with Carol Burnett   SHOW   
2658   ts265844                                           #ABtalks   SHOW   
4274  tm1172010                                  The Lockdown Plan  MOVIE   
4648  tm1113921                                           In Vitro  MOVIE   

                                            description release_year  \
1813  TV legend David Letterman teams up with fascin...   2018-01-01   
1939  MInecraft: Story Mode is an interactive, anima...   2018-01-01   
2386  In this unscripted series starring comedy lege...   2018-01-01   
2658  #ABtalks is a YouTube interview show hosted by...   2018-01-01   
4274                                                NaN   2020-01-01   
4648  'In Vitro' is an otherworldly rumination on me...   2019-01-01   

     age_certification  runtime genres production_countries  seasons  \
1813             TV-MA       50    NaN                   US      4.0   
1939             TV-PG       52    NaN                   US      1.0   
2386              TV-G       24    NaN                   US      1.0   
2658             TV-PG       68    NaN                  NaN      1.0   
4274               NaN       49    NaN                  NaN      NaN   
4648               NaN       27    NaN                  NaN      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1813   tt7829834         7.8      5581.0            8.217         7.6  
1939  tt10498322         5.6       347.0              NaN         NaN  
2386   tt7204366         6.3       237.0            1.621         6.2  
2658  tt12635254         9.6         7.0              NaN         NaN  
4274  tt13079112         6.5         NaN              NaN         NaN  
4648  tt10545994         7.7         NaN              NaN         NaN

Since the core data needed for analysis, `genres`, is missing, we will remove these observations and then check the number of missing values in this column after deletion:

In [21]:
cleaned_titles = cleaned_titles.dropna(subset=["genres"])
cleaned_titles["genres"].isnull().sum()

0

Next, we will assess the missing data in `cleaned_credits`, where only the `character` variable has missing values.

Since the character name does not affect our analysis of actors in high IMDb-rated works across different genres, and the missing values in this variable could be due to the person being a director (with no associated character), we can retain the observations with missing values in the `character` variable.

#### Handling Duplicate Data

Based on the meaning and content of the data variables, there should not be any observations in `cleaned_titles` where all variable values are identical. Therefore, we will check for any duplicate values.

In [22]:
cleaned_titles.duplicated().sum()

0

The output result is 0, indicating that there are no duplicate values.

Next, let's check if there are any duplicate values in the `cleaned_credits` table.

In [23]:
cleaned_credits.duplicated().sum()

0

#### Handling Inconsistent Data

For cleaned_titles, inconsistent data may exist in the `genres` and `production_countries` variables. We will examine whether there are multiple different values referring to the same genre and the same country.

In [24]:
cleaned_titles['genres'].value_counts()

genres
drama            3357
comedy           2419
thriller         1446
action           1339
romance          1080
crime            1066
documentation     981
family            769
animation         732
fantasy           727
european          679
scifi             647
horror            438
history           336
music             266
reality           226
war               221
sport             188
western            53
Name: count, dtype: int64

In [25]:
cleaned_titles['production_countries'].value_counts()

production_countries
US    5648
IN    1610
GB    1068
JP    1046
FR     720
      ... 
GT       1
CU       1
LK       1
NP       1
FO       1
Name: count, Length: 108, dtype: int64

Since there are too many values in the `value_counts` result, Pandas will only display the first and last few values by default. To display the full result, we can set `display.max_rows` to `None`, which removes the row limit for displaying results.

However, since we only need to see the complete result during the current `value_counts` call, we can use `option_context` to temporarily change the limit.

In [26]:
with pd.option_context('display.max_rows', None):
    print(cleaned_titles['production_countries'].value_counts())

production_countries
US         5648
IN         1610
GB         1068
JP         1046
FR          720
KR          637
ES          637
CA          608
DE          383
CN          295
MX          264
IT          224
BR          221
AU          217
TR          195
PH          192
AR          150
ID          149
BE          148
TW          133
NG          131
PL          126
ZA          103
NL          102
HK          102
CO           94
EG           93
DK           89
TH           87
SE           81
LB           70
NO           68
AE           52
IE           49
SG           47
XX           43
IL           42
RU           41
CL           35
CH           33
PS           32
BG           31
MY           30
SA           28
AT           28
IS           28
LU           27
NZ           27
PE           26
RO           25
QA           24
CZ           22
JO           19
FI           18
HU           18
UY           15
MA           15
PT           14
KH           10
KW           10
PR            9
PK 

From the above output, it appears that the countries of origin are represented using two-letter country codes, except for one instance of `Lebanon`.

The country code for `Lebanon` is `LB`, and it appears 70 times, indicating inconsistency in the data. Both `LB` and `Lebanon` refer to the same country and need to be standardized.

In [27]:
cleaned_titles["production_countries"] = cleaned_titles["production_countries"].replace("Lebanon", "LB")

In [28]:
with pd.option_context('display.max_rows', None):
    print(cleaned_titles['production_countries'].value_counts())

production_countries
US    5648
IN    1610
GB    1068
JP    1046
FR     720
KR     637
ES     637
CA     608
DE     383
CN     295
MX     264
IT     224
BR     221
AU     217
TR     195
PH     192
AR     150
ID     149
BE     148
TW     133
NG     131
PL     126
ZA     103
HK     102
NL     102
CO      94
EG      93
DK      89
TH      87
SE      81
LB      71
NO      68
AE      52
IE      49
SG      47
XX      43
IL      42
RU      41
CL      35
CH      33
PS      32
BG      31
MY      30
AT      28
SA      28
IS      28
LU      27
NZ      27
PE      26
RO      25
QA      24
CZ      22
JO      19
FI      18
HU      18
UY      15
MA      15
PT      14
KW      10
KH      10
PK       9
PR       9
UA       8
MT       8
VN       8
LT       7
IR       7
CD       7
SU       7
TN       7
SN       6
AL       6
KE       6
GH       6
IQ       5
MU       5
CY       5
KN       4
GR       4
IO       4
SY       4
TZ       4
MC       4
GL       3
CM       3
AO       3
BS       3
HR       3
BD       3


For `original_credits`, inconsistent data may exist in the `role` field. We will examine whether there are multiple different values referring to the same type of crew member.

In [29]:
cleaned_credits['role'].value_counts()

role
ACTOR       73251
DIRECTOR     4550
Name: count, dtype: int64

From the above output, the `role` field only has two possible values: `ACTOR` or `DIRECTOR`, and there is no inconsistent data. We can convert this column to the `Category` type, which is more memory-efficient than string types and indicates that the values are limited to a small set.

In [31]:
cleaned_credits["role"] = cleaned_credits["role"].astype("category")
cleaned_credits["role"]

0           ACTOR
1           ACTOR
2           ACTOR
3           ACTOR
4           ACTOR
           ...   
77796       ACTOR
77797       ACTOR
77798       ACTOR
77799       ACTOR
77800    DIRECTOR
Name: role, Length: 77801, dtype: category
Categories (2, object): ['ACTOR', 'DIRECTOR']

#### Handling Invalid or Erroneous Data

In [34]:
cleaned_titles.describe()

release_year       runtime      seasons    imdb_score  \
count                          16970  16970.000000  5954.000000  16970.000000   
mean   2015-11-14 22:42:51.974072064     80.912552     2.455492      6.514207   
min              1954-01-01 00:00:00      0.000000     1.000000      1.500000   
25%              2015-01-01 00:00:00     45.000000     1.000000      5.800000   
50%              2018-01-01 00:00:00     90.000000     2.000000      6.600000   
75%              2020-01-01 00:00:00    107.000000     3.000000      7.300000   
max              2022-01-01 00:00:00    225.000000    42.000000      9.500000   
std                              NaN     39.596172     2.869428      1.131095   

         imdb_votes  tmdb_popularity    tmdb_score  
count  1.694100e+04     16842.000000  16515.000000  
mean   3.281655e+04        29.396307      6.846933  
min    5.000000e+00         0.600000      1.000000  
25%    7.800000e+02         4.070000      6.200000  
50%    3.508000e+03        10.195000      6.900000  
75%    1.697800e+04        23.639000      7.500000  
max    2.294231e+06      2274.044000     10.000000  
std    1.141492e+05        93.178235      1.078831

From the above statistics, there are no out-of-context values in `cleaned_titles`.

Since `cleaned_credits` does not contain variables that represent numerical values, there is no need to use describe to check it.

## Data Sorting

In [36]:
cleaned_titles

id                               title   type  \
1       tm84618                         Taxi Driver  MOVIE   
1       tm84618                         Taxi Driver  MOVIE   
2      tm154986                         Deliverance  MOVIE   
2      tm154986                         Deliverance  MOVIE   
2      tm154986                         Deliverance  MOVIE   
...         ...                                 ...    ...   
5846   tm898842                         C/O Kaadhal  MOVIE   
5847  tm1059008                             Lokillo  MOVIE   
5849   ts271048  Mighty Little Bheem: Kite Festival   SHOW   
5849   ts271048  Mighty Little Bheem: Kite Festival   SHOW   
5849   ts271048  Mighty Little Bheem: Kite Festival   SHOW   

                                            description release_year  \
1     A mentally unstable Vietnam War veteran works ...   1976-01-01   
1     A mentally unstable Vietnam War veteran works ...   1976-01-01   
2     Intent on seeing the Cahulawassee River before...   1972-01-01   
2     Intent on seeing the Cahulawassee River before...   1972-01-01   
2     Intent on seeing the Cahulawassee River before...   1972-01-01   
...                                                 ...          ...   
5846  A heart warming film that explores the concept...   2021-01-01   
5847  A controversial TV host and comedian who has b...   2021-01-01   
5849  With winter behind them, Bheem and his townspe...   2021-01-01   
5849  With winter behind them, Bheem and his townspe...   2021-01-01   
5849  With winter behind them, Bheem and his townspe...   2021-01-01   

     age_certification  runtime     genres production_countries  seasons  \
1                    R      114      drama                   US      NaN   
1                    R      114      crime                   US      NaN   
2                    R      109      drama                   US      NaN   
2                    R      109     action                   US      NaN   
2                    R      109   thriller                   US      NaN   
...                ...      ...        ...                  ...      ...   
5846               NaN      134      drama                  NaN      NaN   
5847               NaN       90     comedy                   CO      NaN   
5849               NaN        7     family                  NaN      1.0   
5849               NaN        7  animation                  NaN      1.0   
5849               NaN        7     comedy                  NaN      1.0   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1      tt0075314         8.2    808582.0           40.965       8.179  
1      tt0075314         8.2    808582.0           40.965       8.179  
2      tt0068473         7.7    107673.0           10.010       7.300  
2      tt0068473         7.7    107673.0           10.010       7.300  
2      tt0068473         7.7    107673.0           10.010       7.300  
...          ...         ...         ...              ...         ...  
5846  tt11803618         7.7       348.0              NaN         NaN  
5847  tt14585902         3.8        68.0           26.005       6.300  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  
5849  tt13711094         7.8        18.0            2.289      10.000  

[16970 rows x 15 columns]

In [37]:
cleaned_credits

person_id         id                name                character  \
0          3748    tm84618      Robert De Niro            Travis Bickle   
1         14658    tm84618        Jodie Foster            Iris Steensma   
2          7064    tm84618       Albert Brooks                      Tom   
3          3739    tm84618       Harvey Keitel  Matthew 'Sport' Higgins   
4         48933    tm84618     Cybill Shepherd                    Betsy   
...         ...        ...                 ...                      ...   
77796    736339  tm1059008    Adelaida Buscato                María Paz   
77797    399499  tm1059008  Luz Stella Luengas             Karen Bayona   
77798    373198  tm1059008         Inés Prieto                    Fanny   
77799    378132  tm1059008        Isabel Gaona                   Cacica   
77800   1950416  tm1059008      Julian Gaviria                      NaN   

           role  
0         ACTOR  
1         ACTOR  
2         ACTOR  
3         ACTOR  
4         ACTOR  
...         ...  
77796     ACTOR  
77797     ACTOR  
77798     ACTOR  
77799     ACTOR  
77800  DIRECTOR  

[77801 rows x 5 columns]

The organization of data is closely related to the analysis direction. The goal of this data analysis is to organize films from different genres, such as comedies, action films, and sci-fi films, and calculate the average IMDB rating of the films that actors have appeared in, in order to identify high-rated actors within each genre.

To obtain both genre and actor data simultaneously, we need to join `cleaned_titles` and `cleaned_credits` using `id` as the key, as `id` represents the movie ID in both datasets.

In [38]:
credits_with_titles = pd.merge(cleaned_credits, cleaned_titles, on="id", how="inner")

In [39]:
credits_with_titles.head()

person_id       id            name      character   role        title  \
0      3748  tm84618  Robert De Niro  Travis Bickle  ACTOR  Taxi Driver   
1      3748  tm84618  Robert De Niro  Travis Bickle  ACTOR  Taxi Driver   
2     14658  tm84618    Jodie Foster  Iris Steensma  ACTOR  Taxi Driver   
3     14658  tm84618    Jodie Foster  Iris Steensma  ACTOR  Taxi Driver   
4      7064  tm84618   Albert Brooks            Tom  ACTOR  Taxi Driver   

    type                                        description release_year  \
0  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   
1  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   
2  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   
3  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   
4  MOVIE  A mentally unstable Vietnam War veteran works ...   1976-01-01   

  age_certification  runtime genres production_countries  seasons    imdb_id  \
0                 R      114  drama                   US      NaN  tt0075314   
1                 R      114  crime                   US      NaN  tt0075314   
2                 R      114  drama                   US      NaN  tt0075314   
3                 R      114  crime                   US      NaN  tt0075314   
4                 R      114  drama                   US      NaN  tt0075314   

   imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0         8.2    808582.0           40.965       8.179  
1         8.2    808582.0           40.965       8.179  
2         8.2    808582.0           40.965       8.179  
3         8.2    808582.0           40.965       8.179  
4         8.2    808582.0           40.965       8.179

Since we are only interested in analyzing the reputation of films in which actors have participated, and directors are not within the scope of our analysis, we will filter observations where the `role` is categorized as `ACTOR` for further analysis.

In [40]:
actor_with_titles = credits_with_titles.query('role == "ACTOR"')

In order to identify actors in high IMDB-rated films within each genre, we first need to group by genre and actor.

When grouping actors, we choose to use the `person_id` instead of the `name` variable, as names can be misspelled or duplicated. The cast and crew ID is more accurate in reflecting which specific actor it is.

In [41]:
actor_with_titles.groupby(["genres", "person_id"])["imdb_score"].mean()

genres   person_id
action   1000         6.866667
         100007       7.000000
         100013       6.400000
         100019       6.500000
         100020       6.500000
                        ...   
western  993735       6.500000
         998673       7.300000
         998674       7.300000
         998675       7.300000
         99940        4.000000
Name: imdb_score, Length: 168881, dtype: float64

We can use `reset_index` to reset the hierarchical index and obtain a more organized DataFrame.

In [42]:
imdb_score_groupby_genres_and_person_id_df = actor_with_titles.groupby(["genres", "person_id"])["imdb_score"].mean().reset_index()
imdb_score_groupby_genres_and_person_id_df

genres person_id  imdb_score
0        action      1000    6.866667
1        action    100007    7.000000
2        action    100013    6.400000
3        action    100019    6.500000
4        action    100020    6.500000
...         ...       ...         ...
168876  western    993735    6.500000
168877  western    998673    7.300000
168878  western    998674    7.300000
168879  western    998675    7.300000
168880  western     99940    4.000000

[168881 rows x 3 columns]

The IMDB rating data grouped by genre and actor is now organized, and we can move on to the next steps of the analysis.

However, we can continue with some additional data cleaning, such as grouping the results again to find out the highest average rating of an actor's films in each genre, and the corresponding actor's name with the highest rating.

To achieve this, we need to group by `genres` again, then extract the `imdb_score` variable and calculate its maximum value.

genres
action           9.3
animation        9.3
comedy           9.2
crime            9.5
documentation    9.1
drama            9.5
european         8.9
family           9.3
fantasy          9.3
history          9.1
horror           9.0
music            8.8
reality          8.9
romance          9.2
scifi            9.3
sport            9.1
thriller         9.5
war              8.8
western          8.9
Name: imdb_score, dtype: float64

Once we know the highest score, we can merge the above results with the previously obtained `imdb_score_groupby_genres_and_person_id_df` to identify the actor IDs corresponding to the highest scores. This will reveal which actor achieved the highest average rating.

In [44]:
genres_max_score_with_person_id = pd.merge(imdb_score_groupby_genres_and_person_id_df, genres_max_scores, on=["genres", "imdb_score"])
genres_max_score_with_person_id

genres person_id  imdb_score
0     action     12790         9.3
1     action      1303         9.3
2     action     21033         9.3
3     action    336830         9.3
4     action     86591         9.3
..       ...       ...         ...
131      war    826547         8.8
132  western     22311         8.9
133  western     28166         8.9
134  western     28180         8.9
135  western     93017         8.9

[136 rows x 3 columns]

From the above results, we can see that there may be more than one actor corresponding to the highest score, as multiple actors might have the same average score.

To obtain the actor names corresponding to the actor IDs, we can merge with the `cleaned_credits` DataFrame. Since this DataFrame contains other columns, we only need the `person_id` and `name` columns, so we can first extract these two columns and remove duplicate rows.

In [46]:
actor_id_with_names = cleaned_credits[['person_id', 'name']].drop_duplicates()
actor_id_with_names

person_id                name
0          3748      Robert De Niro
1         14658        Jodie Foster
2          7064       Albert Brooks
3          3739       Harvey Keitel
4         48933     Cybill Shepherd
...         ...                 ...
77796    736339    Adelaida Buscato
77797    399499  Luz Stella Luengas
77798    373198         Inés Prieto
77799    378132        Isabel Gaona
77800   1950416      Julian Gaviria

[54589 rows x 2 columns]

In [50]:
genres_max_score_with_actor_name = pd.merge(genres_max_score_with_person_id, actor_id_with_names, on="person_id")
genres_max_score_with_actor_name

genres person_id  imdb_score                name
0       action     12790         9.3         Olivia Hack
1        scifi     12790         9.3         Olivia Hack
2       action      1303         9.3       Jessie Flower
3    animation      1303         9.3       Jessie Flower
4       family      1303         9.3       Jessie Flower
..         ...       ...         ...                 ...
131        war    826547         8.8         Yuto Uemura
132    western     22311         8.9     Koichi Yamadera
133    western     28166         8.9  Megumi Hayashibara
134    western     28180         8.9      Unsho Ishizuka
135    western     93017         8.9            Aoi Tada

[136 rows x 4 columns]

To group the same genres together, we can use the `sort_values` method to sort the rows in the result by `genres`, and then use `reset_index` to reorder the indices.

After resetting the indices, the DataFrame will have an additional `index` column, which we can then delete.

In [52]:
genres_max_score_with_actor_name = genres_max_score_with_actor_name.sort_values("genres").reset_index().drop("index", axis=1)
genres_max_score_with_actor_name

genres person_id  imdb_score                name
0     action     12790         9.3         Olivia Hack
1     action    336830         9.3     André Sogliuzzo
2     action     21033         9.3          Zach Tyler
3     action     86591         9.3       Cricket Leigh
4     action      1303         9.3       Jessie Flower
..       ...       ...         ...                 ...
131      war    826547         8.8         Yuto Uemura
132  western     28166         8.9  Megumi Hayashibara
133  western     28180         8.9      Unsho Ishizuka
134  western     22311         8.9     Koichi Yamadera
135  western     93017         8.9            Aoi Tada

[136 rows x 4 columns]